rotkehlchen/exchanges/bitfinex.py

```diff
@@ -0,0 +1,933 @@
+            case: Literal['asset_movements'],
+            currency_map: Dict[str, str],
+    ) -> List[AssetMovement]:
+        ...
+
+    def _api_query_paginated(
+            self,
+            options: Dict[str, Any],
+            case: Literal['trades', 'asset_movements'],
+            currency_map: Dict[str, str],
+    ) -> Union[List[Trade], List[AssetMovement]]:
+        """Request a Bitfinex API v2 endpoint paginating via an options
+        attribute.
+
+        Pagination attribute criteria per endpoint:
+          - trades: from `start_ts` to `end_ts` (as offset) with limit 2500.
+          - movements: from `start_ts` to `end_ts` (as offset) with limit 1000.
+
+        Error codes documentation:
+        https://docs.bitfinex.com/docs/abbreviations-glossary#errorinfo-codes
+
+        Rates limit documentation:
+        https://docs.bitfinex.com/docs/requirements-and-limitations#rest-rate-limits
+        """
+        deserialization_method: DeserializationMethod
+        endpoint: Literal['trades', 'movements']
+        response_case: Literal['trades', 'asset_movements']
+        if case == 'trades':
+            endpoint = 'trades'
+            response_case = 'trades'
+            deserialization_method = self._deserialize_trade
+            expected_raw_result_length = API_TRADES_MIN_RESULT_LENGTH
+            id_index = 0
+            timestamp_index = 2
+        elif case == 'asset_movements':
+            endpoint = 'movements'
+            response_case = 'asset_movements'
+            deserialization_method = self._deserialize_asset_movement
+            expected_raw_result_length = API_MOVEMENTS_MIN_RESULT_LENGTH
+            id_index = 0
+            timestamp_index = 5
+        else:
+            raise AssertionError(f'Unexpected {self.name} case: {case}')
+
+        call_options = options.copy()
+        limit = options['limit']
+        results: Union[List[Trade], List[AssetMovement]] = []  # type: ignore # bug list nothing
+        last_result_id = None
+        last_result_timestamp = None
+        retries_left = API_REQUEST_RETRY_TIMES
+        while retries_left >= 0:
+            response = self._api_query(
+                endpoint=endpoint,
+                options=call_options,
+            )
+            if response.status_code != HTTPStatus.OK:
+                # NB: check if the rate limits have been hit (response JSON as dict)
+                try:
+                    response_dict = rlk_jsonloads_dict(response.text)
+                except (AssertionError, JSONDecodeError):
+                    msg = (
+                        f'Error decoding {self.name} {case} unsuccessful response JSON as dict. '
+                        f'Continue decoding unsuccessful response as a JSON list.'
+                    )
+                    log.debug(msg, options=call_options)
+                else:
+                    if response_dict.get('error', None) == API_RATE_LIMITS_ERROR_MESSAGE:
```

Comment:
I did the best I could :(

https://docs.bitfinex.com/docs/requirements-and-limitations#rest-rate-limits
```
If your IP address is rate limited, the API will return the JSON response {“error”: “ERR_RATE_LIMIT”}”.
```
And you can check on the error codes page there is no reference to "limit rate reached".
https://docs.bitfinex.com/docs/abbreviations-glossary#errorinfo-codes

Reply:
Indeed. The only one that may be is `ERR_READY` but we would need it to be confirmed by their support. Perhaps you can ping their support about it. For now this will have to do then. And if their support confirms there is an error code for retry/rate limiting then we can use that.

Reply:
message sent